File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/10-Tree/LeetCode/LeetCode-0637-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        """
        Do a level-by-level traversal using a queue
        We will push the tree nodes into a queue by attaching a "level" value to
        each node. Thus, when we get the node from the queue, we know at which level
        this node resides.
        """
        result = []
        if root == None: return result

        # <node, level>
        Q = []
        currentLevel = 0
        Q.append([root, currentLevel])

        sum = 0   # Sum of the elements at this level
        n = 0     # number of nodes at this level

        while len(Q) > 0:
            p, level = Q.pop(0)

            # Did we go to the next level?
            if level > currentLevel:
                result.append(sum/n)
                sum = n = 0
                currentLevel = level

            n += 1
            sum += p.val

            if p.left != None: Q.append([p.left, level+1])
            if p.right != None: Q.append([p.right, level+1])

        # Push the average for the last level
        if n > 0: result.append(sum/n)
            
        return result
        

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/10-Tree/LeetCode/LeetCode-0637-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    /// Do a level-by-level traversal using a queue
    /// We will push the tree nodes into a queue by attaching a "level" value to
    /// each node. Thus, when we get the node from the queue, we know at which level
    /// this node resides.
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> result;
        if (root == NULL) return result;

        // <node, level>
        queue<pair<TreeNode *, int>> Q;
        int currentLevel = 0;
        Q.push(make_pair(root, currentLevel));

        double sum = 0;   // Sum of the elements at this level
        int n = 0;        // # of nodes at this level

        while (!Q.empty()){
            auto [p, level] = Q.front();
            Q.pop();   // Remove from the queue

            // Did we go to the next level?
            if (level > currentLevel){
                result.push_back(sum/n);
                sum = n = 0;
                currentLevel = level;
            } // end-if

            n++;
            sum += p->val;

            if (p->left) Q.push(make_pair(p->left, level+1));
            if (p->right) Q.push(make_pair(p->right, level+1));
        } //end-while

        // Push the average for the last level
        if (n > 0) result.push_back(sum/n);

        return result;
    } // end-averageOfLevels
};